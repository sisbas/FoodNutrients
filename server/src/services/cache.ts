// Simple in-memory cache placeholder
const cache = new Map<string, any>();

export function getCache<T>(key: string): T | undefined {
  return cache.get(key);
}

export function setCache(key: string, value: any) {
  cache.set(key, value);
}
