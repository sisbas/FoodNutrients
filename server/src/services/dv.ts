export function calculatePercentage(value: number, dailyValue: number): number {
  if (!dailyValue) return 0;
  return value / dailyValue * 100;
}
