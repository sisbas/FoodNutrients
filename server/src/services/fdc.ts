import fetch from "node-fetch";
import { NUTRIENTS_TR, PROFILES } from "@shared";

const SEED_URL = process.env.SEED_JSON_URL!;
const FDC_API_KEY = process.env.FDC_API_KEY;

// Basit in-memory cache (PoC); prod’da Redis/Edge KV kullanın
const memory = new Map<string, any>();

export async function searchFoods(q: string, category: string, lang: string, limit: number) {
  // 1) Seed’ten basit arama
  const seed = await loadSeed();
  const items = seed.filter((x: any) => String(x.description || "").toLowerCase().includes(q.toLowerCase())).slice(0, limit);
  // 2) (Opsiyon) FDC API çağrısı ile zenginleştirme…
  return items.map((x: any) => ({ fdcId: x.fdcId, name_en: x.description }));
}

export async function getFoodDetail(fdcId: string, measure: string, grams: number | undefined, profileId: string) {
  const seed = await loadSeed();
  const item = seed.find((x: any) => String(x.fdcId) === String(fdcId));
  if (!item) throw new Error("Food not found");

  // çok basit normalizasyon (PoC): 100 g değerlerini çıkar
  const per100: Record<string, number> = {};
  for (const n of item.foodNutrients || []) {
    // burada FDC nutrient id → internal key eşleme yapılmalı (NUTRIENTS_TR ile)
    // PoC’ta enerji/protein/yağ/kh gibi alanları tahmini anahtarlarla doldurun.
  }

  const gramsFinal = measure === "custom" && grams ? grams : (measure === "serving" ? 100 : 100); // PoC olarak 100g kabul
  const scaled: Record<string, number> = {};
  Object.entries(per100).forEach(([k, v]) => { scaled[k] = v * gramsFinal / 100; });

  const profile = PROFILES.find(p => p.id === profileId) || PROFILES[0];
  const percentages: Record<string, number> = {};
  Object.entries(scaled).forEach(([k, v]) => {
    const dv = (profile as any).dv[k];
    if (typeof dv === "number" && dv > 0) percentages[k] = v / dv * 100;
  });

  return { fdcId, measure, grams: gramsFinal, per100, values: scaled, percentages, profile: profile.id };
}

async function loadSeed(): Promise<any[]> {
  if (memory.has("seed")) return memory.get("seed");
  const res = await fetch(SEED_URL);
  const json = await res.json();
  memory.set("seed", json);
  return json;
}
