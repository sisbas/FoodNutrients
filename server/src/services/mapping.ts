// Placeholder for FDC â†’ internal model mappings
export function mapNutrientId(id: number): string | undefined {
  // TODO: implement mapping
  return undefined;
}
