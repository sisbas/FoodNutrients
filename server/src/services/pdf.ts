interface Row { label: string; per100: number; value: number; percent: number; }

export function generateHtml(data: { title: string; rows: Row[] }) {
  const rows = data.rows.map(r =>
    `<tr><td>${r.label}</td><td>${r.per100}</td><td>${r.value}</td><td>${r.percent.toFixed(0)}%</td></tr>`
  ).join("");
  return `<!DOCTYPE html><html><head><meta charset="utf-8"/><title>${data.title}</title></head><body><h1>${data.title}</h1><table>${rows}</table></body></html>`;
}
