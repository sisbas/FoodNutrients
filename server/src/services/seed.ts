import fetch from "node-fetch";

// Placeholder for seed JSON loader
export async function loadSeedFromUrl(url: string): Promise<any[]> {
  const res = await fetch(url);
  return res.json();
}
