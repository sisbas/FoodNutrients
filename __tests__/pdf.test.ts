import { generateHtml } from '@server/services/pdf';
import { describe, it, expect } from 'vitest';

describe('pdf html', () => {
  it('matches snapshot', () => {
    const html = generateHtml({
      title: 'Sample',
      rows: [{ label: 'Protein', per100: 5, value: 10, percent: 20 }]
    });
    expect(html).toMatchSnapshot();
  });
});
