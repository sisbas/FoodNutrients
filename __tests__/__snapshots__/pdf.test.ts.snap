exports[`pdf html matches snapshot 1`] = `<!DOCTYPE html><html><head><meta charset="utf-8"/><title>Sample</title></head><body><h1>Sample</h1><table><tr><td>Protein</td><td>5</td><td>10</td><td>20%</td></tr></table></body></html>`;
